File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    private int expand(String s, int left, int right){
        int n = s.length();
        while(left >= 0 && right < n && s.charAt(left) == s.charAt(right)){
            left--; right++;
        }
        return right - left - 1;
    }

    public String longestPalindrome(String s) {
        int start = 0, maxLen = 0, n = s.length();
        for(int i = 0; i< n; i++){
            int oddLen = expand(s, i, i);
            int evenLen = expand(s, i, i+1);
            int len = Math.max(oddLen, evenLen);
            if(len > maxLen){
                maxLen = len;
                start = i - (maxLen - 1)/2;
            }
        }
        return s.substring(start, start + maxLen);
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int longest = 0, n = s.length();
        string res = "";
        
        // Expand from each index
        for(int i = 0; i < n; i++){
            // Check Odd Length
            int left = i, right = i;
            while(left >= 0 && right < n && s[left] == s[right]){
                if(right - left + 1 > longest){
                    longest = right - left + 1;
                    res = s.substr(left, longest);
                }
                left--, right ++;
            }

            // Check Even Length
            left = i, right = i + 1;
            while(left >= 0 && right < n && s[left] == s[right]){
                if(right - left + 1 > longest){
                    longest = right - left + 1;
                    res = s.substr(left, longest);
                }
                left--, right++;
            }
        }

        return res;
    }
};